File: system_control/libraries/file_loader/test_files/kb_go/postgres/kb_construct/kb_construct_module/ConstructStatusTable.go
package kb_construct_module

import (
	"database/sql"
	//"encoding/json"
	"fmt"

	_ "github.com/lib/pq"
)

// ConstructStatusTable manages status table operations
type ConstructStatusTable struct {
	conn        *sql.DB
	constructKB *ConstructKB // Reference to ConstructKB instance
	database    string
	tableName   string
}

// StatusFieldResult represents the result of adding a status field
type StatusFieldResult struct {
	Status     string                 `json:"status"`
	Message    string                 `json:"message"`
	Properties map[string]interface{} `json:"properties"`
	Data       map[string]interface{} `json:"data"`
}

// CheckInstallationResult represents the result of check installation
type CheckInstallationResult struct {
	MissingPathsAdded       int `json:"missing_paths_added"`
	NotSpecifiedPathsRemoved int `json:"not_specified_paths_removed"`
}

// NewConstructStatusTable creates a new instance of ConstructStatusTable
func NewConstructStatusTable(conn *sql.DB, constructKB *ConstructKB, database string) (*ConstructStatusTable, error) {
	cst := &ConstructStatusTable{
		conn:        conn,
		constructKB: constructKB,
		database:    database,
		tableName:   database + "_status",
	}

	fmt.Printf("database: %s\n", database)

	if err := cst.setupSchema(); err != nil {
		return nil, fmt.Errorf("error setting up schema: %w", err)
	}

	return cst, nil
}

// setupSchema sets up the database schema
func (cst *ConstructStatusTable) setupSchema() error {
	// Create ltree extension
	if _, err := cst.conn.Exec("CREATE EXTENSION IF NOT EXISTS ltree;"); err != nil {
		return fmt.Errorf("error creating ltree extension: %w", err)
	}

	// Drop existing table
	dropQuery := fmt.Sprintf("DROP TABLE IF EXISTS %s CASCADE", cst.tableName)
	if _, err := cst.conn.Exec(dropQuery); err != nil {
		return fmt.Errorf("error dropping table: %w", err)
	}

	// Create the status table
	createTableQuery := fmt.Sprintf(`
		CREATE TABLE %s (
			id SERIAL PRIMARY KEY,
			data JSON,
			path LTREE UNIQUE
		);`, cst.tableName)

	if _, err := cst.conn.Exec(createTableQuery); err != nil {
		return fmt.Errorf("error creating table: %w", err)
	}

	// Create indexes
	indexes := []string{
		// GIST index for ltree path operations
		fmt.Sprintf("CREATE INDEX IF NOT EXISTS idx_%s_path_gist ON %s USING GIST (path);",
			cst.tableName, cst.tableName),

		// B-tree index on path for exact lookups
		fmt.Sprintf("CREATE INDEX IF NOT EXISTS idx_%s_path_btree ON %s (path);",
			cst.tableName, cst.tableName),
	}

	for _, indexQuery := range indexes {
		if _, err := cst.conn.Exec(indexQuery); err != nil {
			return fmt.Errorf("error creating index: %w", err)
		}
	}

	fmt.Printf("Status table '%s' created with optimized indexes.\n", cst.tableName)
	return nil
}

// AddStatusField adds a new status field to the knowledge base
func (cst *ConstructStatusTable) AddStatusField(statusKey string, properties map[string]interface{}, description string, initialData map[string]interface{}) (*StatusFieldResult, error) {
	// Type validation is implicit in Go's type system
	// but we can still check for nil maps
	if properties == nil {
		properties = make(map[string]interface{})
	}
	if initialData == nil {
		initialData = make(map[string]interface{})
	}

	fmt.Printf("Added status field '%s' with properties: %v and data: %v\n", statusKey, properties, initialData)

	// Add the node to the knowledge base
	if err := cst.constructKB.AddInfoNode("KB_STATUS_FIELD", statusKey, properties, initialData, description); err != nil {
		return nil, fmt.Errorf("error adding info node: %w", err)
	}

	result := &StatusFieldResult{
		Status:     "success",
		Message:    fmt.Sprintf("Status field '%s' added successfully", statusKey),
		Properties: properties,
		Data:       initialData,
	}

	return result, nil
}

// CheckInstallation synchronizes the knowledge_base and status_table based on paths
func (cst *ConstructStatusTable) CheckInstallation() (*CheckInstallationResult, error) {
	// Get all paths from status_table
	getPathsQuery := fmt.Sprintf("SELECT path FROM %s;", cst.tableName)
	rows, err := cst.conn.Query(getPathsQuery)
	if err != nil {
		return nil, fmt.Errorf("error querying status table paths: %w", err)
	}
	defer rows.Close()

	var allPaths []string
	for rows.Next() {
		var path string
		if err := rows.Scan(&path); err != nil {
			return nil, fmt.Errorf("error scanning path: %w", err)
		}
		allPaths = append(allPaths, path)
	}

	// Get specified paths (paths with label "KB_STATUS_FIELD") from knowledge_table
	specifiedPathsQuery := fmt.Sprintf(`
		SELECT path FROM %s 
		WHERE label = 'KB_STATUS_FIELD';`, cst.database)

	rows, err = cst.conn.Query(specifiedPathsQuery)
	if err != nil {
		return nil, fmt.Errorf("error querying knowledge base paths: %w", err)
	}
	defer rows.Close()

	var specifiedPaths []string
	for rows.Next() {
		var path string
		if err := rows.Scan(&path); err != nil {
			return nil, fmt.Errorf("error scanning specified path: %w", err)
		}
		specifiedPaths = append(specifiedPaths, path)
	}

	fmt.Printf("specified_paths: %v\n", specifiedPaths)

	// Find missing paths (in specified_paths but not in all_paths)
	missingPaths := findDifference(specifiedPaths, allPaths)
	fmt.Printf("missing_paths: %v\n", missingPaths)

	// Find not specified paths (in all_paths but not in specified_paths)
	notSpecifiedPaths := findDifference(allPaths, specifiedPaths)
	fmt.Printf("not_specified_paths: %v\n", notSpecifiedPaths)

	// Begin transaction for consistency
	tx, err := cst.conn.Begin()
	if err != nil {
		return nil, fmt.Errorf("error beginning transaction: %w", err)
	}
	defer tx.Rollback()

	// Process not_specified_paths: remove entries from status_table
	deleteQuery := fmt.Sprintf("DELETE FROM %s WHERE path = $1;", cst.tableName)
	deleteStmt, err := tx.Prepare(deleteQuery)
	if err != nil {
		return nil, fmt.Errorf("error preparing delete statement: %w", err)
	}
	defer deleteStmt.Close()

	for _, path := range notSpecifiedPaths {
		fmt.Printf("deleting path: %s\n", path)
		if _, err := deleteStmt.Exec(path); err != nil {
			return nil, fmt.Errorf("error deleting path %s: %w", path, err)
		}
	}

	// Process missing_paths: add entries to status_table
	insertQuery := fmt.Sprintf(`
		INSERT INTO %s (data, path)
		VALUES ($1, $2);`, cst.tableName)
	insertStmt, err := tx.Prepare(insertQuery)
	if err != nil {
		return nil, fmt.Errorf("error preparing insert statement: %w", err)
	}
	defer insertStmt.Close()

	for _, path := range missingPaths {
		fmt.Printf("inserting path: %s\n", path)
		if _, err := insertStmt.Exec("{}", path); err != nil {
			return nil, fmt.Errorf("error inserting path %s: %w", path, err)
		}
	}

	// Commit the transaction
	if err := tx.Commit(); err != nil {
		return nil, fmt.Errorf("error committing transaction: %w", err)
	}

	result := &CheckInstallationResult{
		MissingPathsAdded:        len(missingPaths),
		NotSpecifiedPathsRemoved: len(notSpecifiedPaths),
	}

	return result, nil
}

File: luajit_programs_and_containers/building_blocks/knowledge_base/postgres/construct_kb/construct_status_table.lua
--[[
  construct_status_table.lua
  
  LuaJIT translation of construct_status_table.py
  Status table construction and synchronization with knowledge base.
  
  Dependencies:
    - DBI (luadbi)
    - dkjson
  
  Usage:
    local ConstructStatusTable = require("construct_status_table")
    local st = ConstructStatusTable.new(conn, construct_kb, database)
]]

local json = require("dkjson")

local Construct_Status_Table = {}
Construct_Status_Table.__index = Construct_Status_Table

local function quote_ident(name)
  return '"' .. name:gsub('"', '""') .. '"'
end

local function quote_literal(val)
  if val == nil then return "NULL" end
  return "'" .. tostring(val):gsub("'", "''") .. "'"
end

---------------------------------------------------------------------------
-- Constructor
---------------------------------------------------------------------------

--- Create a new Construct_Status_Table.
-- @param conn        DBI connection object
-- @param construct_kb  Construct_KB instance
-- @param database    string  base table name
-- @param upload_flag boolean (default false)
function Construct_Status_Table.new(conn, construct_kb, database, upload_flag)
  local self = setmetatable({}, Construct_Status_Table)
  self.conn = conn
  self.construct_kb = construct_kb
  self.database = database
  self.table_name = database .. "_status"
  self.upload_flag = upload_flag or false

  print("database: " .. self.database)

  if not self.upload_flag then
    self:_setup_schema()
  end

  return self
end

---------------------------------------------------------------------------
-- Internal helpers
---------------------------------------------------------------------------

function Construct_Status_Table:_exec(sql_str)
  local stmt, err = self.conn:prepare(sql_str)
  if not stmt then
    error("SQL prepare error: " .. tostring(err) .. "\nQuery: " .. sql_str)
  end
  local ok, exec_err = stmt:execute()
  if not ok then
    error("SQL execute error: " .. tostring(exec_err) .. "\nQuery: " .. sql_str)
  end
  stmt:close()
end

function Construct_Status_Table:_query(sql_str)
  local stmt, err = self.conn:prepare(sql_str)
  if not stmt then
    error("SQL prepare error: " .. tostring(err) .. "\nQuery: " .. sql_str)
  end
  local ok, exec_err = stmt:execute()
  if not ok then
    error("SQL execute error: " .. tostring(exec_err) .. "\nQuery: " .. sql_str)
  end
  local rows = {}
  local row = stmt:fetch(true)
  while row do
    rows[#rows + 1] = row
    row = stmt:fetch(true)
  end
  stmt:close()
  return rows
end

---------------------------------------------------------------------------
-- Schema setup
---------------------------------------------------------------------------

function Construct_Status_Table:_setup_schema()
  local tn = self.table_name

  self:_exec("CREATE EXTENSION IF NOT EXISTS ltree;")
  self:_exec(string.format("DROP TABLE IF EXISTS %s CASCADE;", quote_ident(tn)))

  self:_exec(string.format([[
    CREATE TABLE %s (
      id SERIAL PRIMARY KEY,
      data JSON,
      path LTREE UNIQUE
    )
  ]], quote_ident(tn)))

  -- Indexes
  self:_exec(string.format("CREATE INDEX IF NOT EXISTS %s ON %s USING GIST (path)",
    quote_ident("idx_" .. tn .. "_path_gist"), quote_ident(tn)))
  self:_exec(string.format("CREATE INDEX IF NOT EXISTS %s ON %s (path)",
    quote_ident("idx_" .. tn .. "_path_btree"), quote_ident(tn)))

  print("Status table '" .. tn .. "' created with optimized indexes.")
end

---------------------------------------------------------------------------
-- Public API
---------------------------------------------------------------------------

--- Add a status field to the knowledge base.
-- @param status_key string
-- @param properties table
-- @param description string
-- @param initial_data table
-- @return table  result summary
function Construct_Status_Table:add_status_field(status_key, properties, description, initial_data)
  assert(type(status_key) == "string", "status_key must be a string")
  assert(type(description) == "string", "description must be a string")
  assert(type(initial_data) == "table", "initial_data must be a table")

  if properties == nil then
    properties = {}
  end
  assert(type(properties) == "table", "properties must be a table")

  if os.getenv("DCS_KB_VERBOSE") == "1" then
    print("Added status field '" .. status_key ..
          "' with properties: " .. json.encode(properties) ..
          " and data: " .. json.encode(initial_data))
  end

  self.construct_kb:add_info_node("KB_STATUS_FIELD", status_key, properties, initial_data, description)

  return {
    status = "success",
    message = "Status field '" .. status_key .. "' added successfully",
    properties = properties,
    data = initial_data,
  }
end

--- Synchronize status table with knowledge base.
-- @return table  summary of changes
function Construct_Status_Table:check_installation()
  local tn = quote_ident(self.table_name)
  local db = quote_ident(self.database)

  -- Get all paths from status table
  local status_rows = self:_query(string.format("SELECT path FROM %s;", tn))
  local all_paths = {}
  local all_paths_set = {}
  for _, row in ipairs(status_rows) do
    local p = tostring(row.path)
    all_paths[#all_paths + 1] = p
    all_paths_set[p] = true
  end

  -- Get specified paths from knowledge base
  local kb_rows = self:_query(string.format([[
    SELECT path FROM %s WHERE label = 'KB_STATUS_FIELD'
  ]], db))

  local specified_paths = {}
  local specified_set = {}
  for _, row in ipairs(kb_rows) do
    local p = tostring(row.path)
    specified_paths[#specified_paths + 1] = p
    specified_set[p] = true
  end
  local VERBOSE = os.getenv("DCS_KB_VERBOSE") == "1"
  if VERBOSE then print("specified_paths: " .. table.concat(specified_paths, ", ")) end

  -- Find missing paths (in specified but not in status table)
  local missing_paths = {}
  for _, path in ipairs(specified_paths) do
    if not all_paths_set[path] then
      missing_paths[#missing_paths + 1] = path
    end
  end
  if VERBOSE then print("missing_paths: " .. table.concat(missing_paths, ", ")) end

  -- Find not-specified paths (in status table but not in knowledge base)
  local not_specified_paths = {}
  for _, path in ipairs(all_paths) do
    if not specified_set[path] then
      not_specified_paths[#not_specified_paths + 1] = path
    end
  end
  if VERBOSE then print("not_specified_paths: " .. table.concat(not_specified_paths, ", ")) end

  -- Delete not-specified entries (batched; PG supports long IN lists).
  if #not_specified_paths > 0 then
    local lits = {}
    for _, path in ipairs(not_specified_paths) do
      lits[#lits + 1] = quote_literal(path)
      if VERBOSE then print("deleting path: " .. path) end
    end
    -- Chunk to avoid pathologically long statements.
    local CHUNK = 500
    for i = 1, #lits, CHUNK do
      local j = math.min(i + CHUNK - 1, #lits)
      self:_exec(string.format("DELETE FROM %s WHERE path IN (%s);",
        tn, table.concat(lits, ",", i, j)))
    end
  end

  -- Insert missing entries as a single multi-row VALUES (chunked).
  if #missing_paths > 0 then
    local tuples = {}
    for _, path in ipairs(missing_paths) do
      tuples[#tuples + 1] = "('{}'," .. quote_literal(path) .. ")"
      if VERBOSE then print("inserting path: " .. path) end
    end
    local CHUNK = 500
    for i = 1, #tuples, CHUNK do
      local j = math.min(i + CHUNK - 1, #tuples)
      self:_exec(string.format("INSERT INTO %s (data, path) VALUES %s",
        tn, table.concat(tuples, ",", i, j)))
    end
  end

  return {
    missing_paths_added = #missing_paths,
    not_specified_paths_removed = #not_specified_paths,
  }
end

return Construct_Status_Table

